File: HISTORY - CODE/Comatose v1.2/bin/pyComa/BOSS.py
from __future__ import division
from __future__ import absolute_import
from symbol import symbol, arrow
import board
import lvlchange
import sys
import time
from creature import creature, creatureR
import setlvl

class BOSS(board.board):
    def __init__(self, cr, inn):
        super(BOSS, self).__init__(None, False, inn) #super(board, self) #False for vertical
        #crhv? wBOSS? finBOSS?
        self.crBOSS = cr
    def outsideBOSS(self):
        for k in xrange(board.FIN):
            self.Boarder(k, symbol.E)
        for i in xrange(board.WIDTH):
            self.Boarder(i, symbol.BD)
        for i in xrange(board.FIN - 1, board.FIN - board.WIDTH, -1):
            self.Boarder(i, symbol.BD)
        for i in xrange(0, board.FIN, board.WIDTH):
            self.Boarder(i, symbol.BD)
        for i in xrange(board.WIDTH - 1, board.FIN + 1, board.WIDTH):
            self.Boarder(i, symbol.BD)
    def printcrackBOSS(self):
        if not board.creature.opening:
            for cr in self.crBOSS:
                self.Boarder(cr, symbol.CE)
        else:
            for cr in self.crBOSS:
                self.Boarder(cr, symbol.E)
    def printboardBOSS(self, P, H, B, CCRthis):
        self.outsideBOSS()
        self.printcrackBOSS()
        self.F(self)

        B.bombs(P.pickingup(B.bombs(), H.loc()))
        P.printp(self.Boarder())
        B.printb(self)
        self.printcBOSS(CCRthis, B)
        if H.printh():
            self.seeboard(B.bombs())
        else:
            sys.stdout.write(u"Fredrick does not wake up.\n\r")
            return True
        if not lvlchange.fast:
            time.sleep(0.250) #250ms
        time.sleep(0.250) #
        return False
    def nxtlvlBOSS(self, H):
        for cr in self.crBOSS:
            if H.nextlvll(cr, self.crachv):
                return True
        return False
    def cBOSSloc(self, C, _type):
        u''' cBOSSloc
        * - location of creatures (start)
        * - direction of creatures (start)
        '''
        t = int(board.WIDTH / 2)
        u = int(board.HEIGHT / 2)
        m = u * board.WIDTH + t
        C.append(_type(m, arrow.U))
    def printcBOSS(self, CCRthis, B):
        CCthis = self.CCthis
        class DoubleIter(object):
            def __init__(self):
                if len(CCthis) != len(CCRthis):
                    raise ValueError
                self.count = 0
            def __iter__(self):
                while self.count < len(CCthis):
                    yield (self.count, CCthis[self.count], CCRthis[self.count])
                    self.count += 1
        for count, aCC, aCCR in DoubleIter():
            aCC.moving()
            aCCR.moving()

            opennings = False
            opennings = self.openningBOSS(aCC.loc, B)
            if not opennings:
                opennings = self.openningBOSS(aCCR.loc, B)

            for cr in self.crBOSS:
                aCC.cbomb(cr, self.crachv, B.loc)
                aCCR.cbomb(cr, self.crachv, B.loc)
            board.creature.opening = opennings

            if not aCC.loc or not aCCR.loc:
                del CCthis[count]
                del CCRthis[count]

            self.Boarder(aCC.loc, symbol.C)
            self.Boarder(aCCR.loc, symbol.C)

        setlvl.nc += 1
        if setlvl.nc > setlvl.timing:
            setlvl.nc = 0
            if setlvl.TOTAL > len(self.CCthis):
                self.cBOSSloc(self.CCthis, creature)
            if setlvl.TOTAL > len(CCRthis):
                self.cBOSSloc(CCRthis, creatureR)

    def completee(self):
        for cr in self.crBOSS:
            if self.Boarder(cr + self.crachv) != symbol.B or self.Boarder(cr - self.crachv) != symbol.B:
                return False
        return True
    def clandd(self, loc):
        for cr in self.crBOSS:
            if loc == cr + self.crachv or loc == cr - self.crachv:
                return True
        return False
    def both(self, ret, BB, loc, v):
        if ret:
            for cr in self.crBOSS:
                A = cr + self.crachv
                B = cr - self.crachv
                C = self.crBOSS[8] - v
                D = self.crBOSS[7] + v
                for k,ll in enumerate(BB.loc):
                    if ll == A or ll == B or ll == C or ll == D:
                        if ll != loc:
                            del BB.loc[k]
        return ret
    def openningBOSS(self, loc, BB):
        v = 0
        if self.crachv == 1:
            v = board.WIDTH
        elif self.crachv == board.WIDTH:
            v = 1
        cland = False
        complete = True
        complete = self.completee()
        if complete and (self.Boarder(self.crBOSS[8] - v) != symbol.B or self.Boarder(self.crBOSS[7] + v) != symbol.B):
            complete = False
        cland = self.clandd(loc)
        if not cland and (loc == self.crBOSS[8] - v or loc == self.crBOSS[7] + v):
            cland = True
        return self.both(cland and complete, BB, loc, v)
    def levelBOSS(self, P, H, B, CCRthis):
        while not lvlchange.done:
            if not self.printboardBOSS(P, H, B, CCRthis):
                if self.nxtlvlBOSS(H):
                    lvlchange.done = True
            else:
                lvlchange.done = True
                lvlchange.gameover = True
    def calllvlBOSS(self, P, H, B, CCRthis):
        if not lvlchange.gameover:
            lvl = lvlchange.Thread(target=self.levelBOSS, args=(P, H, B, CCRthis))
            lvl.start()
            lvlchange.hmove(H, B)
            lvl.join()
        return lvlchange.gameover

File: src/C++/BOSS.cpp
#include "BOSS.h"

const MAX BOSS::wBOSS;
const MAX BOSS::finBOSS;

BOSS::BOSS(MAX* cr, INSIDE in)
	: board(in, wBOSS), crBOSS(cr)
{
}

BOSS::~BOSS()
{
}

void BOSS::outsideBOSS()
{
	for (CNT k = 0; k < fin; k++)
	{
		Boarder(k, symbol::E);
	}

	for (CNT i = 0; i < wBOSS; i++)
	{
		Boarder(i, symbol::BD);
	}
	for (CNT i = finBOSS - 1; i > finBOSS - wBOSS; i--)
	{
		Boarder(i, symbol::BD);
	}
	for (CNT i = 0; i < finBOSS; i += wBOSS)
	{
		Boarder(i, symbol::BD);
	}
	for (CNT i = wBOSS - 1; i <= finBOSS; i += wBOSS)
	{
		Boarder(i, symbol::BD);
	}
}

void BOSS::printcrackBOSS()
{
	if (!creature::opening())
	{
		for (CNT i = 0; i < 9; i++)
		{
			Boarder(crBOSS[i], symbol::CE);
		}
	}
	else
	{
		for (CNT i = 0; i < 9; i++)
		{
			Boarder(crBOSS[i], symbol::E);
		}
	}
}

bool BOSS::printboardBOSS(pickup& P, CNT& bombs, human& H, bomb& B, vector<creatureR>* CCRthis, CNT& nc, MAX& timing, MAX& tot)
{
	outsideBOSS();
	printcrackBOSS();
	F(wBOSS, h, *this);

	P.pickingup(*B.bombs, *H.locOFh);
	P.printp(Boarder());
	B.printb(*this);
	printcBOSS(CCRthis, B, nc, timing, tot);
	if (H.printh((*Boarder()), wBOSS))
	{
		seeboard(*B.bombs);
	}
	else
	{
		cout << "Fredrick does not wake up." << endl;
		return true;
	}

	if (!lvlchange::fast()) Sleep(250);
	Sleep(250); //
	return false;
}

bool BOSS::nxtlvlBOSS()
{
	for (CNT i = 0; i < 9; i++)
	{
		if (human::nextlvll(crBOSS[i], crachv))
		{
			return true;
		}
	}
	return false;
}

/* cBOSSloc
* - location of creatures (start)
* - direction of creatures (start)
*/
void BOSS::cBOSSloc(vector<creature>& C)
{
	MAX t = wBOSS / 2;
	MAX u = h / 2;
	MAX m = u * wBOSS + t;
	C.push_back(creature(m, arrow::U, Boarder()));
}

void BOSS::cBOSSloc(vector<creatureR>& CR)
{
	MAX t = wBOSS / 2;
	MAX u = h / 2;
	MAX m = u * wBOSS + t;
	CR.push_back(creatureR(m, arrow::U, Boarder()));
}

void BOSS::printcBOSS(vector<creatureR>* CCRthis, bomb& B, CNT& nc, MAX& timing, const MAX& total)
{
	MAX size = CCthis->size();
	for (CNT j = 0; j < size; j++)
	{
		CCthis->at(j).moving(wBOSS);
		CCRthis->at(j).moving(wBOSS);

		bool opennings = false;
		opennings = openningBOSS(CCthis->at(j).loc, B);
		if (!opennings) opennings = openningBOSS(CCRthis->at(j).loc, B);
		if (opennings) creature::finopening();

		for (CNT f = 0; f < 9; f++)
		{
			CCthis->at(j).cbomb(w, crBOSS[f], crachv, B.loc());
			CCRthis->at(j).cbomb(w, crBOSS[f], crachv, B.loc());
		}

		if (!CCthis->at(j).loc || !CCRthis->at(j).loc)
		{
			vector<creature>::iterator it;
			vector<creatureR>::iterator itR;
			it = CCthis->begin();
			itR = CCRthis->begin();
			it += j;
			itR += j;
			if (CCthis->end() > it)
				CCthis->erase(it);
			if (CCRthis->end() > itR)
				CCRthis->erase(itR);
			if (0 != j) j--;
		}

		Boarder(CCthis->at(j).loc, symbol::C);
		Boarder(CCRthis->at(j).loc, symbol::C);
		size = CCthis->size();
	}

	nc++;
	if (nc > timing)
	{
		nc = 0;
		if (total > CCthis->size()) cBOSSloc(*CCthis);
		if (total > CCRthis->size()) cBOSSloc(*CCRthis);
	}
}

bool BOSS::completee()
{
	for (CNT j = 0; j < 9; j++)
	{
		if (Boarder(crBOSS[j] + crachv) != symbol::B || Boarder(crBOSS[j] - crachv) != symbol::B)
		{
			return false;
		}
	}
	return true;
}

bool BOSS::clandd(whole& loc)
{
	for (CNT j = 0; j < 9; j++)
	{
		if (loc == crBOSS[j] + crachv || loc == crBOSS[j] - crachv)
		{
			return true;
		}
	}
	return false;
}

bool BOSS::both(bool ret, bomb& BB, whole& loc, MAX& v)
{
	if (ret)
	{
		for (whole j = 0; j < 9; j++)
		{
			whole A = crBOSS[j] + crachv;
			whole B = crBOSS[j] - crachv;
			whole C = crBOSS[8] - v;
			whole D = crBOSS[7] + v;
			whole size = BB.loc()->size();
			for (CNT k = 0; k < size; k++)
			{
				whole ll = BB.loc()->at(k);
				if (ll == A || ll == B || ll == C || ll == D)
				{
					if (ll != loc)
					{
						BB.loc()->erase(BB.loc()->begin() + k);
						k--;
					}
				}
				size = BB.loc()->size();
			}
		}
	}
	return ret;
}

bool BOSS::openningBOSS(whole loc, bomb& BB)
{
	MAX v = 0;
	if (crachv == 1) v = wBOSS;
	else if (crachv == wBOSS) v = 1;

	bool cland = false, complete = true;

	complete = completee();
	if (complete && (Boarder(crBOSS[8] - v) != symbol::B || Boarder(crBOSS[7] + v) != symbol::B))
	{
		complete = false;
	}

	cland = clandd(loc);
	if (!cland && (loc == crBOSS[8] - v || loc == crBOSS[7] + v))
	{
		cland = true;
	}

	return both(cland && complete, BB, loc, v);
}

bool BOSS::calllvlBOSS(pickup& P, CNT& bombs, human& H, bomb& B, vector<creatureR>* CCRthis, CNT& nc, MAX& timing, MAX& tot)
{
	if (!gmover())
	{
		thread lvl(lvlchange::levelBOSS, ref(*this), ref(P), ref(bombs), ref(H), ref(B), CCRthis, ref(nc), ref(timing), ref(tot));
		hmove(H, bombs, B);
		lvl.join();
	}
	return gmover();
}
